lib/pptx/presentation.go

```diff
@@ -0,0 +1,163 @@
+// pptx is a package to create slide presentations in pptx (Microsoft Power Point) format.
+// A `.pptx` file is just a bunch of zip compressed `.xml` files following the Office Open XML (OOXML) format.
+// To see its content, you can just `unzip <path/to/file>.pptx -d <folder>`.
+// With this package, it is possible to create a `Presentation` and add `Slide`s to it.
+// Then, when saving the presentation, it will generate the required `.xml` files, compress them and write to the disk.
+// Note that this isn't a full implementation of the OOXML format, but a wrapper around it.
+// There's a base template with common files to the presentation and then when saving, the package generate only the slides and relationships.
+// The base template and slide templates were generated using https://python-pptx.readthedocs.io/en/latest/
+// For more information about OOXML, check http://officeopenxml.com/index.php
+package pptx
+
+import (
+	"archive/zip"
+	"bytes"
+	_ "embed"
+	"fmt"
+	"image/png"
+	"os"
+)
+
+type Presentation struct {
+	Title       string
+	Description string
+	Subject     string
+	Creator     string
+	D2Version   string
+
+	Slides []*Slide
+}
+
+type Slide struct {
+	BoardPath   []string
+	Image       []byte
+	ImageWidth  int
+	ImageHeight int
+	ImageTop    int
+	ImageLeft   int
+}
+
+func NewPresentation(title, description, subject, creator, d2Version string) *Presentation {
+	return &Presentation{
+		Title:       title,
+		Description: description,
+		Subject:     subject,
+		Creator:     creator,
+		D2Version:   d2Version,
+	}
+}
+
+func (p *Presentation) AddSlide(pngContent []byte, boardPath []string) error {
+	src, err := png.Decode(bytes.NewReader(pngContent))
+	if err != nil {
+		return fmt.Errorf("error decoding PNG image: %v", err)
+	}
+
+	var width, height int
+	srcSize := src.Bounds().Size()
+	srcWidth, srcHeight := float64(srcSize.X), float64(srcSize.Y)
+
+	// compute the size and position to fit the slide
+	// if the image is wider than taller and its aspect ratio is, at least, the same as the available image space aspect ratio
+	// then, set the image width to the available space and compute the height
+	if srcWidth/srcHeight >= IMAGE_ASPECT_RATIO {
+		width = SLIDE_WIDTH
+		height = int(float64(width) * (srcHeight / srcWidth))
+		if height > IMAGE_HEIGHT {
+			// this would overflow with the title, so we need to adjust to use only IMAGE_WIDTH
+			width = IMAGE_WIDTH
+			height = int(float64(width) * (srcHeight / srcWidth))
+		}
+	} else {
+		// otherwise, this image could overflow the slide height/header
+		height = IMAGE_HEIGHT
+		width = int(float64(height) * (srcWidth / srcHeight))
+	}
+	top := (IMAGE_HEIGHT - height) / 2
+	left := (SLIDE_WIDTH - width) / 2
```

Comment:
an ascii art here would be good, not sure why one is image height and the other is slide width